common/script/index.coffee

```diff
@@ -72,13 +72,49 @@ api.daysSince = (yesterday, options = {}) ->
   Math.abs api.startOfDay(_.defaults {now:yesterday}, o).diff(api.startOfDay(_.defaults {now:o.now}, o), 'days')
 
 ###
-  Should the user do this taks on this date, given the task's repeat options and user.preferences.dayStart?
+  Should the user do this task on this date, given the task's repeat options and user.preferences.dayStart?
 ###
-api.shouldDo = (day, repeat, options={}) ->
-  return false unless repeat
+api.shouldDo = (day, dailyTask, options = {}) ->
+  return false unless dailyTask.type == 'daily' && dailyTask.repeat
+  if !dailyTask.startDate
+    dailyTask.startDate = moment().toDate()
+  if dailyTask.startDate instanceof String
+    dailyTask.startDate = moment(dailyTask.startDate).toDate()
   o = sanitizeOptions options
-  selected = repeat[api.dayMapping[api.startOfDay(_.defaults {now:day}, o).day()]]
-  return selected
+  day = api.startOfDay(_.defaults {now:day}, o)
+  dayOfWeekNum = day.day() # e.g. 1 for Monday if week starts on Mon
+
+  # check if event is today or in the future
+  hasStartedCheck = day >= api.startOfDay(_.defaults {now:dailyTask.startDate}, o)
+
+  if dailyTask.frequency == 'daily'
+    daysSinceTaskStart = api.numDaysApart(day, dailyTask.startDate, o)
+    everyXCheck = (daysSinceTaskStart % dailyTask.everyX == 0)
+    return everyXCheck && hasStartedCheck
+  else if dailyTask.frequency == 'weekly'
+    dayOfWeekCheck = dailyTask.repeat[api.dayMapping[dayOfWeekNum]]
+    weeksSinceTaskStartWeek = api.numWeeksApart(day, dailyTask.startDate, o)
+    everyXCheck = (weeksSinceTaskStartWeek % dailyTask.everyX == 0)
+    return dayOfWeekCheck && everyXCheck && hasStartedCheck
+  else
+    # unexpected frequency string
```

Comment:
I think this will cause a problem for any Daily that hasn't been created or edited with the new code (i.e., Dailies created on the current website, or the mobile apps, or third-party apps, or direct API calls). Dailies created with those methods could persist for months after this code goes live. This code needs to be fully backwards-compatible with the old style of Dailies.